Raven.Database/Bundles/Replication/Tasks/IndexReplicationTask.cs

```diff
@@ -44,16 +56,192 @@ public IndexReplicationTask(DocumentDatabase database, HttpRavenRequestFactory h
+                    
+                    foreach (var destination in replicationDestinations)
+                    {
+                        var url = destination.ConnectionStringOptions.Url;
+
+                        var replicatedIndexIds = replicatedSideBySideIndexIds.GetOrAdd(url, new ConcurrentSet<int>());
+                        var indexesToReplicate = sideBySideIndexesToReplicate
+                            .Where(x => replicatedIndexIds.Contains(x.Key) == false).ToList();
+
+                        if (indexesToReplicate.Count == 0)
+                            continue;
+
+                        try
+                        {
+                            ReplicateSideBySideIndexesMultiPut(destination, indexesToReplicate.Select(x => x.Value).ToList());
+                            foreach (var index in indexesToReplicate)
+                            {
+                                replicatedIndexIds.TryAdd(index.Key);
+                            }
+
+                            if (commonReplicatedIndexIds.Count == 0)
+                            {
+                                commonReplicatedIndexIds.UnionWith(replicatedIndexIds);
+                            }
+                            else
+                            {
+                                commonReplicatedIndexIds.IntersectWith(replicatedIndexIds);
+                            }
+                        }
+                        catch (Exception e)
+                        {
+                            failedToReplicate = true;
+                            Log.ErrorException("Failed to replicate side by side index to " + destination, e);
+                        }
+                    }
+
+                    if (failedToReplicate)
+                        return;
+
+                    //since we replicated the side by side indexes for all destinations,
+                    //there won't be any need to replicate them again
+                    foreach (var indexId in commonReplicatedIndexIds)
+                    {
+                        IndexToAdd _;
+                        sideBySideIndexesToReplicate.TryRemove(indexId, out _);
+                    }
+                }
+                catch (Exception e)
+                {
+                    Log.ErrorException("Failed to replicate side by side indexes", e);
+                }
+                finally
+                {
+                    interlockedLock.Exit();
+
+                    var newCount = Interlocked.Read(ref sideBySideWorkCounter);
+                    if (currentWorkCount != newCount)
+                    {
+                        //handling a race condition where we get a new side by side index to replicate
+                        //and exiting while thinking that this index will be handled
+                        ReplicateSideBySideIndexes();
+                    }
+                }
+            }, TaskCreationOptions.LongRunning);
```

Comment:
so?
Consider the common case where you are creating and killing threads

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811



On Mon, Aug 15, 2016 at 2:24 PM, grisha-kotler <notifications@github.com>
wrote:

> In Raven.Database/Bundles/Replication/Tasks/IndexReplicationTask.cs
> <https://github.com/ayende/ravendb/pull/4220#discussion_r74748875>:
>
> > +                {
> > +                    Log.ErrorException("Failed to replicate side by side indexes", e);
> > +                }
> > +                finally
> > +                {
> > +                    interlockedLock.Exit();
> > +
> > +                    var newCount = Interlocked.Read(ref sideBySideWorkCounter);
> > +                    if (currentWorkCount != newCount)
> > +                    {
> > +                        //handling a race condition where we get a new side by side index to replicate
> > +                        //and exiting while thinking that this index will be handled
> > +                        ReplicateSideBySideIndexes();
> > +                    }
> > +                }
> > +            }, TaskCreationOptions.LongRunning);
>
> This can take time when some destinations are offline.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/ayende/ravendb/pull/4220/files/19daed7dd220dda0006b1132c5a9f9edc546f622..5647fd686e34a8090c05bf45b97a5bb4b081bc89#r74748875>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHIs_m6uw2FQso-BSSqd_HJKTpQojV-ks5qgExQgaJpZM4JkSsG>
> .
>